File: backend/app/api/approvals.py
"""Phase 4 — human approval endpoints.

Flow:
  1. An agent submits a GatewayRequest that requires approval. The
     gateway redacts, stores a `pending_ai_approvals` row, and the
     response carries `pending_approval_id` + `approval_preview`.
  2. The phone (or HTML demo page) lists pending items via
     `GET /approvals`, fetches the preview via `GET /approvals/{id}`,
     and the human chooses Approve / Reject.
  3. Approve  → `POST /approvals/{id}/approve` runs the queued external
     call with the EXACT redacted prompt the user just saw.
  4. Reject   → `POST /approvals/{id}/reject` updates the audit log
     and the queued row; nothing is ever sent.

The redacted prompt and redaction map (placeholder→category only) are
the only payload bits the user can see — originals never come through
the API.
"""
from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal
from typing import Literal

from fastapi import APIRouter, Depends, HTTPException, Query, Response, status
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.db import get_db
from app.gateway import GatewayPolicyError, GatewayResponse, LocalAIGateway
from app.models.pending_approval import PendingApproval


router = APIRouter(prefix="/approvals", tags=["approvals"])


# -- Response shapes ---------------------------------------------------------

class ApprovalListItem(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    agent_id: str
    request_type: str
    privacy_level: str
    provider: str
    model: str
    estimated_cost_gbp: Decimal
    status: str
    created_at: datetime
    expires_at: datetime


class ApprovalDetail(ApprovalListItem):
    """Full preview for the approval modal — redacted only."""
    redacted_prompt: str
    redaction_map: dict[str, str]


# -- Endpoints ---------------------------------------------------------------

@router.get("", response_model=list[ApprovalListItem])
def list_approvals(
    user_id: uuid.UUID = Query(..., description="Owner whose pending approvals to list"),
    status_filter: Literal["pending", "approved", "rejected", "expired", "executed"] | None
        = Query(None, alias="status"),
    db: Session = Depends(get_db),
) -> list[ApprovalListItem]:
    stmt = select(PendingApproval).where(PendingApproval.user_id == user_id)
    if status_filter:
        stmt = stmt.where(PendingApproval.status == status_filter)
    stmt = stmt.order_by(PendingApproval.created_at.desc())
    rows = db.execute(stmt).scalars().all()
    return [_to_list_item(r) for r in rows]


@router.get("/{approval_id}", response_model=ApprovalDetail)
def get_approval(
    approval_id: uuid.UUID,
    db: Session = Depends(get_db),
) -> ApprovalDetail:
    row = db.get(PendingApproval, approval_id)
    if row is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "pending approval not found")
    return _to_detail(row)


@router.post("/{approval_id}/approve", response_model=GatewayResponse)
async def approve(
    approval_id: uuid.UUID,
    db: Session = Depends(get_db),
) -> GatewayResponse:
    gateway = LocalAIGateway(db=db)
    try:
        return await gateway.execute_approved(approval_id)
    except GatewayPolicyError as exc:
        raise HTTPException(status.HTTP_409_CONFLICT, str(exc))


@router.post("/{approval_id}/reject")
def reject(
    approval_id: uuid.UUID,
    db: Session = Depends(get_db),
) -> Response:
    gateway = LocalAIGateway(db=db)
    try:
        gateway.reject_pending(approval_id)
    except GatewayPolicyError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, str(exc))
    return Response(status_code=status.HTTP_204_NO_CONTENT)


# -- Helpers -----------------------------------------------------------------

def _to_list_item(row: PendingApproval) -> ApprovalListItem:
    return ApprovalListItem(
        id=row.id,
        user_id=row.user_id,
        agent_id=row.agent_id,
        request_type=row.request_type,
        privacy_level=row.privacy_level,
        provider=row.provider,
        model=row.model,
        estimated_cost_gbp=row.estimated_cost_gbp,
        status=row.status,
        created_at=row.created_at,
        expires_at=row.expires_at,
    )


def _to_detail(row: PendingApproval) -> ApprovalDetail:
    return ApprovalDetail(
        **_to_list_item(row).model_dump(),
        redacted_prompt=row.redacted_prompt,
        redaction_map=dict(row.redaction_map or {}),
    )

File: mobile/src/api/approvals.ts
/**
 * Approval API client. Mirrors the FastAPI endpoints in
 * `backend/app/api/approvals.py`.
 *
 * No auth headers yet (JWT lands later). All requests go through the
 * gateway-protected backend; this file only deals with the redacted
 * approval queue.
 */
import { API_BASE_URL } from "../config/api";

export type ApprovalStatus =
  | "pending"
  | "approved"
  | "rejected"
  | "expired"
  | "executed";

export interface ApprovalListItem {
  id: string;
  agent_id: string;
  request_type: string;
  privacy_level: string;
  provider: string;
  model: string;
  status: ApprovalStatus;
  estimated_cost_gbp: string;
  expires_at: string;
  created_at: string;
}

export interface ApprovalDetail extends ApprovalListItem {
  redacted_prompt: string;
  redaction_map: Record<string, string>;
  max_output_tokens: number;
}

export interface GatewayResponse {
  audit_log_id: string;
  final_status: string;
  provider: string;
  model: string;
  external_call: boolean;
  text?: string | null;
  error?: string | null;
  input_tokens: number;
  output_tokens: number;
  estimated_cost_gbp: string;
}

async function handle<T>(resp: Response): Promise<T> {
  if (!resp.ok) {
    let detail = `${resp.status} ${resp.statusText}`;
    try {
      const body = await resp.json();
      if (body?.detail) detail = `${resp.status}: ${body.detail}`;
    } catch {
      /* non-JSON error body */
    }
    throw new Error(detail);
  }
  // 204s have no body.
  if (resp.status === 204) return undefined as unknown as T;
  return (await resp.json()) as T;
}

export async function listApprovals(
  userId: string,
  status: ApprovalStatus = "pending"
): Promise<ApprovalListItem[]> {
  const url = `${API_BASE_URL}/approvals?user_id=${encodeURIComponent(
    userId
  )}&status=${status}`;
  const resp = await fetch(url);
  return handle<ApprovalListItem[]>(resp);
}

export async function getApproval(id: string): Promise<ApprovalDetail> {
  const resp = await fetch(`${API_BASE_URL}/approvals/${id}`);
  return handle<ApprovalDetail>(resp);
}

export async function approveApproval(id: string): Promise<GatewayResponse> {
  const resp = await fetch(`${API_BASE_URL}/approvals/${id}/approve`, {
    method: "POST",
  });
  return handle<GatewayResponse>(resp);
}

export async function rejectApproval(id: string): Promise<void> {
  const resp = await fetch(`${API_BASE_URL}/approvals/${id}/reject`, {
    method: "POST",
  });
  await handle<void>(resp);
}
